Do not validate disabled fields

haacked opened the pull request:
Fixes #90 

Note this does not handle disabled fieldsets because we just rely on the browser for that. If that assumption is incorrect, we can fix that in a subsequent PR.

haacked committed:
Add simple disabled input demo

haacked committed:
Do not validate disabled inputs

Fixes #90

dahlbyk commented on src/index.ts:
```diff
@@ -1286,6 +1286,15 @@ export class ValidationService {
+     * @param input
+     * @returns
+     */
+    private isDisabled(input: HTMLElement) {
+        return input.getAttribute('disabled') !== null;
```
Using `ValidatableElement` should let you access the various inputs' `disabled` property. I'm pretty sure `getAttribute()` wouldn't correctly account for disabled `fieldset`, but the property might?

```suggestion
    private isDisabled(input: ValidatableElement) {
        return input.disabled;
```

haacked committed:
Use ValidatableElement instead of HTMLElement

Co-authored-by: Keith Dahlby <dahlbyk@gmail.com>

haacked commented:
I noticed that when a disabled element is valid, we mark it green. We should completely ignore disabled elements. I'll dig into this a bit more.

haacked committed:
Do not track disabled inputs

haacked committed:
Update Disabled Inputs demo to scan for changes

haacked committed:
Minor tidying up

haacked committed:
Reset field when disabled.

haacked commented:
Ok, I updated the disabled fields demo to test dynamic support for fields that are enabled/disabled. Turns out the value of `disabled` is not in the attributes collection nor in the `oldValue` property of `MutationRecord`, so we need to do some special casing.

Note: for this demo to work, we need to be watching for DOM changes, hence:

```html
<script>
    const service = new aspnetValidation.ValidationService(console);
    service.bootstrap({ watch: true });
</script>
```

You can dynamically enable/disable the two inputs and validation works correctly. For example, by default, input 1 is disabled and input 2 is enabled, so submitting the form with no values yields:

<img width="623" alt="image" src="https://github.com/haacked/aspnet-client-validation/assets/19977/6a4c287f-2211-41bf-80f2-b96a2a307d7f">

If I enable input 1 and then try to submit again, we see it now shows an error message.

<img width="593" alt="Screenshot 2024-01-05 at 7 00 16 PM" src="https://github.com/haacked/aspnet-client-validation/assets/19977/e1582bbe-2a60-49bb-b144-e36c3a6a1824">

If I disable both, the form is valid and submits successfully.

dahlbyk commented on src/index.ts:
```diff
@@ -1077,7 +1085,7 @@ export class ValidationService {
         for (let i = 0; i < inputs.length; i++) {
             let input = inputs[i];
-            if (remove) {
+            if (remove || input.disabled) {
```
Hmm… I'm not thrilled with needing to depend on `watch` to validate an input that has been un-`disabled`. That would be an unexpected deviation from jQuery Validation.

I think instead we want it to keep it wired up but make sure validation of a disabled field never fails and won't show a message or summary entry.

dahlbyk commented on src/index.ts:
```diff
@@ -1404,15 +1421,30 @@ export class ValidationService {
+                    const target = mutation.target as ValidatableElement;
+                    if (target.disabled) {
+                        this.remove(target);
+                        this.resetField(this.getElementUID(target));
```
I don't think we want to `remove()`, but I do like the observer resetting validation messages if we notice it becomes disabled. 👍

haacked replied:
I believe the remove is necessary otherwise we highlight the field in green when it is valid.

<img width="548" alt="image" src="https://github.com/haacked/aspnet-client-validation/assets/19977/948ee2ae-b63d-4eb1-b75f-f9312dbc72e1">

Let me look into that though. Maybe I just need to prevent that code from running when the field is disabled too.

dahlbyk reviewed (commented):
We ended up not using `watch: true`. I think we should try to make this work as expected without.

The demo would re-validate the input in `updateStatus()` to handle the reset - if a form has enabled/disable behavior I don't think that's too much to ask.

haacked commented:
> We ended up not using `watch: true`. I think we should try to make this work as expected without.
> 
> The demo would re-validate the input in `updateStatus()` to handle the reset - if a form has enabled/disable behavior I don't think that's too much to ask.

I think we should have two demos. It should Just Work™️ if `watch: true`. Otherwise, we make the user re-validate the input if they disable/enable it.

haacked committed:
Add example of handling disabled/enabled with no watch

haacked committed:
Add a new reset method for resetting a single form input

This can be called after changing the state of a form input.

haacked commented:
Ok, try that out and let me know what you think.

haacked commented:
@dahlbyk do you want to take another look or should I go ahead and merge?

dahlbyk reviewed (commented):
Well I apparently had some pending comments, so let's start there? 😅

haacked commented on src/index.ts:
```diff
@@ -1077,7 +1097,7 @@ export class ValidationService {
         for (let i = 0; i < inputs.length; i++) {
             let input = inputs[i];
-            if (remove) {
+            if (remove || this.isDisabled(input)) {
```
If the field becomes un-disabled, then there are one of two possibilities:

1. If they're using `watch: true`, we call `this.reset(target);` on the field and validation is enabled for the field.
2. If they're not using `watch: true` then they need to call `service.reset(input);` as seen in `DisabledInputsNoWatch.cshtml` line 99.

dahlbyk replied:
Right. My concern is that jQuery Validate will Just Work™️ for fields that become un-disabled, without user action.

We shouldn't pretend disabled inputs don't exist - we should register them like normal and check `isDisabled()` each time we validate.

haacked replied:
Ok, I made this change.

haacked committed:
Always track disabled fields

This way, if they become enabled, we'll validate them. If they become disabled, we won't validate them.

haacked commented:
I made some changes thanks to @dahlbyk's feedback. His feedback made me realize that ignoring validation on a disabled field is orthogonal to whether we update the input's current validation state when it is enabled/disabled.

So we never validate a disabled input. We always validate an enabled input.

However, if an input is currently enabled and showing a validation error and you dynamically disable the input, the validation error will continue to be displayed. But the next time you submit the form, the field will be ignored and the validation message will be cleared.

If you want the validation message to be cleared at the time the input is disabled, there's two options:

1. `service.bootstrap({ watch: true });`
2. `service.reset(input);`

I've included demos of both.

haacked commented on src/index.ts:
```diff
@@ -1296,14 +1316,25 @@ export class ValidationService {
+    private isDisabled(input: Element) {
+        // Test the the input is validatable and disabled
+        const validatableElement = input as ValidatableElement;
+        return validatableElement && validatableElement.disabled;
```
Ok, so perhaps this is better?

```js
private isDisabled(input: Element) {
    // If the input is not validatable, then the `disabled` property is undefined and this returns false, which is what we want.
    return (input as ValidatableElement).disabled;
}
```

haacked committed:
Make the code clearer

`as` in TypeScript is not the same as in C#.

haacked committed:
Add the script files

dahlbyk reviewed (approved):
I didn't kick the tires, but the changes align with what I had in mind. 👍

lonix1 commented:
There's no docs for the resetting bits, so for anyone who lands here after searching the repo:

> If you want the validation message to be cleared at the time the input is disabled, there's two options:
>    service.bootstrap({ watch: true });
>   service.reset(input);

I've found that `service.removeError` works for me, but `service.reset` does not (but sometimes does, weird). Unsure of the difference between the two, but the former works as expected.
